# Capstone 2: Erasmus Program Mobility 

Launched in 1987, the Erasmus program (named after Desiderius Erasmus of Rotterdam, a leading Renaissance figure of European intellectual life) is a European Union program that supports education, training, youth and sport in Europe. The program has 27 EU Member States and 6 non-EU associated countries. It's objective is to promote transnational learning mobility and cooperation between organisations and policy-makers.




## DATA WRANGLING

## Table of Contents
* [Problem Identification](#problem_identification)
* [Data Collection](#data_collection)
* [Data Merging](#data_merging)
* [Data Cleaning](#data_cleaning)
    *  [Review and Handle Missing Data](#missing_data)   
    *  [Normalize Categorical Data](#categorical_data)
    *  [Convert Data Types](#convert_data)
    *  [Review Date Columns](#review_dates)
    *  [Remove Duplicates](#remove_duplicates) 
    *  [Rename Columns](#rename_columns) 


## Problem Identification <a class="anchor" id="problem_identification"></a>

Erasmus funds projects in various countries. Based on past funding from 2014 to 2019, analyze previous patterns and factors to identify trends and predict which countries are most likely to receive future allocations.

## Data Collection <a id="data_collection"></a>

The primary data, sourced from the European Union's official portal for European data, spans over three million entries illustrating individual projects that began between the years 2014-2019: https://data.europa.eu/data/datasets. 

Ten additional files provide funding detail: https://erasmus-plus.ec.europa.eu/projects/projects-lists.

In [1]:
import os
import pandas as pd
import numpy as np
import glob
import re

In [2]:
# Load the projects dataset
projects = pd.read_csv('C:/Users/midol/Documents/Springboard/Springboard/Capstone_2/Erasmus/archive (2)/Erasmus_mobility_statistics_2014_2019.gzip', compression='gzip')

C:\Users\midol\AppData\Local\Temp\ipykernel_13464\497316016.py:2: DtypeWarning: Columns (14) have mixed types. Specify dtype option on import or set low_memory=False.
  projects = pd.read_csv('C:/Users/midol/Documents/Springboard/Springboard/Capstone_2/Erasmus/archive (2)/Erasmus_mobility_statistics_2014_2019.gzip', compression='gzip')


In [3]:
projects.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3462268 entries, 0 to 3462267
Data columns (total 24 columns):
 #   Column                               Dtype  
---  ------                               -----  
 0   Project Reference                    object 
 1   Academic Year                        object 
 2   Mobility Start Month                 object 
 3   Mobility End Month                   object 
 4   Mobility Duration                    float64
 5   Activity (mob)                       object 
 6   Field of Education                   object 
 7   Participant Nationality              object 
 8   Education Level                      object 
 9   Participant Gender                   object 
 10  Participant Profile                  object 
 11  Special Needs                        object 
 12  Fewer Opportunities                  object 
 13  GroupLeader                          object 
 14  Participant Age                      object 
 15  Sending Country Code            

In [4]:
#Load the funding datasets
folder_path = 'C:/Users/midol/Documents/Springboard/Springboard/Capstone_2/Erasmus/Funding files'
file_list = glob.glob(folder_path + "/*.csv") 
funding = pd.DataFrame(pd.read_csv(file_list[0])) 

for i in range(1,len(file_list)): 
    data = pd.read_csv(file_list[i]) 
    df = pd.DataFrame(data) 
    funding = pd.concat([funding,df],ignore_index=True) 

In [5]:
funding.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 134114 entries, 0 to 134113
Data columns (total 5 columns):
 #   Column                   Non-Null Count   Dtype 
---  ------                   --------------   ----- 
 0   Key Action               134113 non-null  object
 1   Action Type              134113 non-null  object
 2   Call year                134113 non-null  object
 3   Project Identifier       134111 non-null  object
 4   EU Grant award in euros  134110 non-null  object
dtypes: object(5)
memory usage: 5.1+ MB


## Data Merging <a id="data_merging"></a>

In [6]:
merged = pd.merge(projects, funding, how='inner', left_on='Project Reference', right_on='Project Identifier')

In [7]:
merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3462106 entries, 0 to 3462105
Data columns (total 29 columns):
 #   Column                               Dtype  
---  ------                               -----  
 0   Project Reference                    object 
 1   Academic Year                        object 
 2   Mobility Start Month                 object 
 3   Mobility End Month                   object 
 4   Mobility Duration                    float64
 5   Activity (mob)                       object 
 6   Field of Education                   object 
 7   Participant Nationality              object 
 8   Education Level                      object 
 9   Participant Gender                   object 
 10  Participant Profile                  object 
 11  Special Needs                        object 
 12  Fewer Opportunities                  object 
 13  GroupLeader                          object 
 14  Participant Age                      object 
 15  Sending Country Code            

## Data Cleaning <a id="data_cleaning"></a>

In [8]:
# Rename column headers to the conventional lower case with underscore.
merged = merged.rename(columns=lambda x: x.lower().replace(' ', '_'))

In [9]:
# Display the first few rows of the dataset
merged.head().T

0  \
project_reference                            2014-1-AT01-KA101-000059   
academic_year                                               2014-2015   
mobility_start_month                                          2014-07   
mobility_end_month                                            2014-07   
mobility_duration                                                12.0   
activity_(mob)                                  Staff training abroad   
field_of_education                               Language acquisition   
participant_nationality                                            AT   
education_level                                     ??? - ? Unknown ?   
participant_gender                                             Female   
participant_profile                                             Staff   
special_needs                                                      No   
fewer_opportunities                                                No   
groupleader                                                        No   
participant_age                                                    42   
sending_country_code                                               AT   
sending_city                                                     Wels   
sending_organization                          BRG Wels Wallererstraße   
sending_organisation_erasmus_code                                   -   
receiving_country_code                                             ES   
receiving_city                                                 Malaga   
receiving_organization                       Escuela de Idiomas Nerja   
receiving_organisation_erasmus_code                                 -   
participants                                                      1.0   
key_action                           Learning Mobility of Individuals   
action_type                           School education staff mobility   
call_year                                                        2014   
project_identifier                           2014-1-AT01-KA101-000059   
eu_grant_award_in_euros                                       17205.0   

                                                                    1  \
project_reference                            2014-1-AT01-KA101-000059   
academic_year                                               2014-2015   
mobility_start_month                                          2014-07   
mobility_end_month                                            2014-08   
mobility_duration                                                 7.0   
activity_(mob)                                  Staff training abroad   
field_of_education                               Language acquisition   
participant_nationality                                            AT   
education_level                                     ??? - ? Unknown ?   
participant_gender                                             Female   
participant_profile                                             Staff   
special_needs                                                      No   
fewer_opportunities                                                No   
groupleader                                                        No   
participant_age                                                    25   
sending_country_code                                               AT   
sending_city                                                     Wels   
sending_organization                          BRG Wels Wallererstraße   
sending_organisation_erasmus_code                                   -   
receiving_country_code                                             UK   
receiving_city                                             Canterbury   
receiving_organization                                   Pilgrims LTD   
receiving_organisation_erasmus_code                                 -   
participants                                                      1.0   
key_action                           Learning Mobili

### Review and Handle Missing Data <a id="missing_data"></a>

In [10]:
# Count the number of missing values in each column with %.
missing = pd.concat([merged.isnull().sum(), 100 * merged.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing.sort_values(by='count', ascending=False)

count         %
sending_country_code                   107  0.003091
participant_nationality                106  0.003062
receiving_country_code                  79  0.002282
sending_city                            64  0.001849
participants                            39  0.001126
receiving_organisation_erasmus_code     39  0.001126
receiving_city                          39  0.001126
receiving_organization                  34  0.000982
project_reference                       30  0.000867
groupleader                             30  0.000867
project_identifier                      30  0.000867
sending_organisation_erasmus_code       30  0.000867
sending_organization                    30  0.000867
participant_age                         30  0.000867
fewer_opportunities                     30  0.000867
special_needs                           30  0.000867
participant_profile                     30  0.000867
participant_gender                      30  0.000867
education_level                         30  0.000867
field_of_education                      30  0.000867
activity_(mob)                          30  0.000867
mobility_duration                       30  0.000867
mobility_end_month                      24  0.000693
eu_grant_award_in_euros                 12  0.000347
key_action                              10  0.000289
action_type                             10  0.000289
call_year                               10  0.000289
mobility_start_month                     3  0.000087
academic_year                            0  0.000000

To impute `sending_country_code`, we review information in `sending_organization`.

In [11]:
# Review 'sending_organization' where 'sending_country_code' is NaN.
nan_send_country = merged[merged['sending_country_code'].isna()]
nan_send_country['sending_organization'].value_counts(dropna=False)

sending_organization
UNIVERSITY OF NAMIBIA                           36
NAMIBIA UNIVERSITY OF SCIENCE AND TECHNOLOGY    34
NaN                                             30
POLYTECHNIC OF NAMIBIA PON                       7
Name: count, dtype: int64

Namibia's country code can be used to impute NaN `sending_country_code` rows that have a `sending_organization` in Namibia.

In [12]:
# Impute 'sending_country_code' using country code of 'sending_organization'. 
merged.loc[merged['sending_organization'].eq('UNIVERSITY OF NAMIBIA'), 'sending_country_code'] = 'NA'
merged.loc[merged['sending_organization'].eq('NAMIBIA UNIVERSITY OF SCIENCE AND TECHNOLOGY'), 'sending_country_code'] = 'NA'
merged.loc[merged['sending_organization'].eq('POLYTECHNIC OF NAMIBIA PON'), 'sending_country_code'] = 'NA'

We review the remaining NaN `sending_country_code` rows.

In [13]:
merged[merged['sending_country_code'].isna()]

project_reference              academic_year mobility_start_month  \
3297740               NaN                          -                    1   
3297741               NaN                          -                    1   
3297742               NaN                          -                    1   
3297743               NaN                          -                    1   
3297744               NaN                          -                    1   
3297745               NaN                          -                    1   
3297746               NaN                          -                    1   
3297747               NaN                          -                    1   
3297748               NaN                          -                    1   
3297749               NaN                          -                    1   
3297750               NaN                          -                    1   
3297751               NaN                          -                    1   
3297752               NaN                          -                    1   
3297753               NaN                          -                    1   
3297754               NaN                          -                    1   
3297755               NaN                          -                    1   
3297756               NaN                          -                    1   
3297757               NaN                          -                    1   
3297758               NaN  Water service corporation                  NaN   
3297759               NaN  Water service corporation                  NaN   
3297760               NaN  Water service corporation                  NaN   
3297761               NaN                          -                    1   
3297762               NaN                          -                    1   
3297763               NaN                          -                    1   
3297764               NaN           THE AV WAREHOUSE                    -   
3297765               NaN           THE AV WAREHOUSE                    -   
3297766               NaN           THE AV WAREHOUSE                    -   
3297767               NaN               DOLMEN HOTEL                    -   
3297768               NaN               DOLMEN HOTEL                    -   
3297769               NaN               DOLMEN HOTEL                    -   

        mobility_end_month  mobility_duration activity_(mob)  \
3297740                NaN                NaN            NaN   
3297741                NaN                NaN            NaN   
3297742                NaN                NaN            NaN   
3297743                NaN                NaN            NaN   
3297744                NaN                NaN            NaN   
3297745                NaN                NaN            NaN   
3297746                NaN                NaN            NaN   
3297747                NaN                NaN            NaN   
3297748                NaN                NaN            NaN   
3297749                NaN                NaN            NaN   
3297750                NaN                NaN            NaN   
3297751                NaN                NaN            NaN   
3297752                NaN                NaN            NaN   
3297753                NaN                NaN            NaN   
3297754                NaN                NaN            NaN   
3297755                NaN                NaN            NaN   
3297756                NaN                NaN            NaN   
3297757                NaN                NaN            NaN   
3297758                NaN                NaN            NaN   
3297759                NaN                NaN            NaN   
3297760                NaN                NaN            NaN   
3297761                NaN                NaN            NaN   
3297762                NaN                NaN            NaN   
3297763                NaN                NaN            NaN   
3297764             

The remaining NaN `sending_country_code` rows contain no useful data so these are removed.

In [14]:
# Drop remaining NaN rows.
merged = merged.dropna(subset=['sending_country_code'])

In [15]:
# Review remaining missing values.
missing = pd.concat([merged.isnull().sum(), 100 * merged.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing.sort_values(by='count', ascending=False)

count         %
participant_nationality                 76  0.002195
receiving_country_code                  49  0.001415
sending_city                            34  0.000982
participants                             9  0.000260
receiving_organisation_erasmus_code      9  0.000260
receiving_city                           9  0.000260
receiving_organization                   4  0.000116
eu_grant_award_in_euros                  2  0.000058
sending_country_code                     0  0.000000
project_identifier                       0  0.000000
call_year                                0  0.000000
action_type                              0  0.000000
key_action                               0  0.000000
sending_organisation_erasmus_code        0  0.000000
sending_organization                     0  0.000000
project_reference                        0  0.000000
academic_year                            0  0.000000
groupleader                              0  0.000000
fewer_opportunities                      0  0.000000
special_needs                            0  0.000000
participant_profile                      0  0.000000
participant_gender                       0  0.000000
education_level                          0  0.000000
field_of_education                       0  0.000000
activity_(mob)                           0  0.000000
mobility_duration                        0  0.000000
mobility_end_month                       0  0.000000
mobility_start_month                     0  0.000000
participant_age                          0  0.000000

Next, we see that 92% of `participant_nationality` in this dataset match the corresponding `sending_country_code`.

In [16]:
# Calculate the percentage of rows where 'participant_nationality' code matches the 'sending_country_code'.
diff = (merged['participant_nationality'] == merged['sending_country_code'])
(len(merged[diff] == True) / len(merged)) * 100

92.06282588828206

We therefore use the corresponding `sending_country_code` to impute `participant_nationality`.

In [17]:
# Replace NaN values in 'participant_nationality' with values from 'sending_country_code'.
merged['participant_nationality'] = np.where(merged['participant_nationality'].isna(), merged['sending_country_code'], merged['participant_nationality'])

In [18]:
# Review remaining missing values.
missing = pd.concat([merged.isnull().sum(), 100 * merged.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing.sort_values(by='count', ascending=False)

count         %
receiving_country_code                  49  0.001415
sending_city                            34  0.000982
participants                             9  0.000260
receiving_organisation_erasmus_code      9  0.000260
receiving_city                           9  0.000260
receiving_organization                   4  0.000116
eu_grant_award_in_euros                  2  0.000058
sending_country_code                     0  0.000000
project_identifier                       0  0.000000
call_year                                0  0.000000
action_type                              0  0.000000
key_action                               0  0.000000
sending_organisation_erasmus_code        0  0.000000
sending_organization                     0  0.000000
project_reference                        0  0.000000
academic_year                            0  0.000000
groupleader                              0  0.000000
fewer_opportunities                      0  0.000000
special_needs                            0  0.000000
participant_profile                      0  0.000000
participant_gender                       0  0.000000
education_level                          0  0.000000
participant_nationality                  0  0.000000
field_of_education                       0  0.000000
activity_(mob)                           0  0.000000
mobility_duration                        0  0.000000
mobility_end_month                       0  0.000000
mobility_start_month                     0  0.000000
participant_age                          0  0.000000

To impute `receiving_country_code`, we review information in `receiving_organization`.

In [19]:
# Review 'receiving_organization' where 'receiving_country_code' is NaN.
nan_receive_country = merged[merged['receiving_country_code'].isna()]
nan_receive_country['receiving_organization'].value_counts(dropna=False)

receiving_organization
NAMIBIA UNIVERSITY OF SCIENCE AND TECHNOLOGY    29
UNIVERSITY OF NAMIBIA                           15
POLYTECHNIC OF NAMIBIA PON                       5
Name: count, dtype: int64

We can impute `receiving_country_code` using the location of the `receiving_organization`.

In [20]:
# Impute 'receiving_country_code' using country code of 'receiving_organization'. 
merged.loc[merged['receiving_organization'].eq('UNIVERSITY OF NAMIBIA'), 'receiving_country_code'] = 'NA'
merged.loc[merged['receiving_organization'].eq('NAMIBIA UNIVERSITY OF SCIENCE AND TECHNOLOGY'), 'receiving_country_code'] = 'NA'
merged.loc[merged['receiving_organization'].eq('POLYTECHNIC OF NAMIBIA PON'), 'receiving_country_code'] = 'NA'

In [21]:
# Review remaining missing values.
missing = pd.concat([merged.isnull().sum(), 100 * merged.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing.sort_values(by='count', ascending=False)

count         %
sending_city                            34  0.000982
participants                             9  0.000260
receiving_organisation_erasmus_code      9  0.000260
receiving_city                           9  0.000260
receiving_organization                   4  0.000116
eu_grant_award_in_euros                  2  0.000058
sending_country_code                     0  0.000000
project_identifier                       0  0.000000
call_year                                0  0.000000
action_type                              0  0.000000
key_action                               0  0.000000
receiving_country_code                   0  0.000000
sending_organisation_erasmus_code        0  0.000000
sending_organization                     0  0.000000
project_reference                        0  0.000000
academic_year                            0  0.000000
groupleader                              0  0.000000
fewer_opportunities                      0  0.000000
special_needs                            0  0.000000
participant_profile                      0  0.000000
participant_gender                       0  0.000000
education_level                          0  0.000000
participant_nationality                  0  0.000000
field_of_education                       0  0.000000
activity_(mob)                           0  0.000000
mobility_duration                        0  0.000000
mobility_end_month                       0  0.000000
mobility_start_month                     0  0.000000
participant_age                          0  0.000000

Next, to impute `sending_city`, we review information in `sending_organization`.

In [22]:
# Review 'sending_organization' where 'sending_city' is NaN.
nan_sending_city = merged[merged['sending_city'].isna()]
nan_sending_city['sending_organization'].value_counts(dropna=False)

sending_organization
Scoil Anagaire                             26
UWC Sverige                                 6
COMMISSIONE SINODALE PER LA DIACONIA        1
PECO - PROGETTI EUROPEI DI COOPERAZIONE     1
Name: count, dtype: int64

To impute `sending_city`, we use the city of the `sending_organization`, easily obtained through an online search.

In [23]:
merged.loc[merged['sending_organization'].eq('Scoil Anagaire'), 'sending_city'] = 'Annagry'
merged.loc[merged['sending_organization'].eq('UWC Sverige'), 'sending_city'] = 'Stockholm'
merged.loc[merged['sending_organization'].eq('COMMISSIONE SINODALE PER LA DIACONIA'), 'sending_city'] = 'Luserna San Giovanni'
merged.loc[merged['sending_organization'].eq('PECO - PROGETTI EUROPEI DI COOPERAZIONE'), 'sending_city'] = 'Genoa'

In [24]:
# Review remaining missing values.
missing = pd.concat([merged.isnull().sum(), 100 * merged.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing.sort_values(by='count', ascending=False)

count         %
participants                             9  0.000260
receiving_organisation_erasmus_code      9  0.000260
receiving_city                           9  0.000260
receiving_organization                   4  0.000116
eu_grant_award_in_euros                  2  0.000058
sending_country_code                     0  0.000000
project_identifier                       0  0.000000
call_year                                0  0.000000
action_type                              0  0.000000
key_action                               0  0.000000
receiving_country_code                   0  0.000000
sending_organisation_erasmus_code        0  0.000000
sending_organization                     0  0.000000
sending_city                             0  0.000000
project_reference                        0  0.000000
academic_year                            0  0.000000
groupleader                              0  0.000000
fewer_opportunities                      0  0.000000
special_needs                            0  0.000000
participant_profile                      0  0.000000
participant_gender                       0  0.000000
education_level                          0  0.000000
participant_nationality                  0  0.000000
field_of_education                       0  0.000000
activity_(mob)                           0  0.000000
mobility_duration                        0  0.000000
mobility_end_month                       0  0.000000
mobility_start_month                     0  0.000000
participant_age                          0  0.000000

Next, `participants` refers to the number of participants per project. 

In [25]:
# Review NaN.
merged[merged['participants'].isna()]

project_reference academic_year mobility_start_month  \
3297720  2019-1-FR01-KA102-060462     2019-2020              2019-09   
3297721  2019-1-FR01-KA102-060462     2019-2020              2019-09   
3297722  2019-1-FR01-KA102-060462     2019-2020              2019-09   
3297725  2019-1-FR01-KA102-060462     2019-2020              2019-09   
3297727  2019-1-FR01-KA102-060462     2019-2020              2019-09   
3297737  2019-1-FR01-KA102-060462     2019-2020              2019-10   
3298466  2019-1-FR01-KA102-060939     2019-2020              2019-09   
3298550  2019-1-FR01-KA102-060939     2019-2020              2019-10   
3298556  2019-1-FR01-KA102-060939     2019-2020              2019-10   

        mobility_end_month  mobility_duration  \
3297720            2019-10                2.0   
3297721            2019-10                2.0   
3297722            2019-10               19.0   
3297725            2019-10               19.0   
3297727            2019-10               19.0   
3297737            2019-10                2.0   
3298466            2019-12               90.0   
3298550            2020-01               90.0   
3298556            2020-01               90.0   

                                            activity_(mob)  \
3297720  Mobility of VET learners (2 weeks up to 3 months)   
3297721  Mobility of VET learners (2 weeks up to 3 months)   
3297722  Mobility of VET learners (2 weeks up to 3 months)   
3297725  Mobility of VET learners (2 weeks up to 3 months)   
3297727  Mobility of VET learners (2 weeks up to 3 months)   
3297737  Mobility of VET learners (2 weeks up to 3 months)   
3298466  ErasmusPro - Mobility of VET learners (3 to 12...   
3298550  ErasmusPro - Mobility of VET learners (3 to 12...   
3298556  ErasmusPro - Mobility of VET learners (3 to 12...   

                            field_of_education participant_nationality  \
3297720                              Education                      FR   
3297721                              Education                      FR   
3297722                      Personal services                      FR   
3297725                      Personal services                      FR   
3297727                      Personal services                      FR   
3297737                              Education                      FR   
3298466  Environment, not elsewhere classified                      FR   
3298550           Manufacturing and processing                      FR   
3298556                                 Sports                      FR   

                                           education_level participant_gender  \
3297720  ISCED-6 - First cycle / Bachelor’s or equivale...             Female   
3297721  ISCED-7 - Second cycle / Master’s or equivalen...               Male   
3297722                ISCED-3 - Upper secondary education             Female   
3297725                ISCED-3 - Upper secondary education             Female   
3297727                ISCED-3 - Upper secondary education               Male   
3297737  ISCED-7 - Second cycle / Master’s or equivalen...             Female   
3298466  ISCED-7 - Second cycle / Master’s or equivalen...             Female   
3298550    ISCED-4 - Post-secondary non-tertiary education               Male   
3298556  ISCED-6 - First cycle / Bachelor’s or equivale...             Female   

         ... receiving_country_code receiving_city  \
3297720  ...                     BE      Bruxelles   
3297721  ...                     BE      Bruxelles   
3297722  ...                     BE      Bruxelles   
3297725  ...                     BE      Bruxelles   
3297727  ...                     BE      Bruxelles   
3297737  ...                     BE      Bruxelles   
3298466  ...                     MT       Pembroke   
3298550  ...                     MT     Haz-Zebbug   
3298556  ...                     MT   St Pauls Bay   

                         receiving_organization  \
3297720  INSTITUT ROYAL POUR SOURDS ET 

We will impute the mean. 

In [26]:
# Impute mean.
part_mean = round(merged['participants'].mean())
print(part_mean)
merged['participants'].fillna(value=part_mean, inplace=True)

1


In [27]:
# Review remaining missing values.
missing = pd.concat([merged.isnull().sum(), 100 * merged.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing.sort_values(by='count', ascending=False)

count         %
receiving_organisation_erasmus_code      9  0.000260
receiving_city                           9  0.000260
receiving_organization                   4  0.000116
eu_grant_award_in_euros                  2  0.000058
sending_country_code                     0  0.000000
project_identifier                       0  0.000000
call_year                                0  0.000000
action_type                              0  0.000000
key_action                               0  0.000000
participants                             0  0.000000
receiving_country_code                   0  0.000000
sending_organisation_erasmus_code        0  0.000000
sending_organization                     0  0.000000
sending_city                             0  0.000000
project_reference                        0  0.000000
academic_year                            0  0.000000
groupleader                              0  0.000000
fewer_opportunities                      0  0.000000
special_needs                            0  0.000000
participant_profile                      0  0.000000
participant_gender                       0  0.000000
education_level                          0  0.000000
participant_nationality                  0  0.000000
field_of_education                       0  0.000000
activity_(mob)                           0  0.000000
mobility_duration                        0  0.000000
mobility_end_month                       0  0.000000
mobility_start_month                     0  0.000000
participant_age                          0  0.000000

To impute `receiving_city`, we review information in `receiving_organization`.

In [28]:
# Review 'receiving_organization' where 'receiving_city' is NaN.
nan_receiving_city = merged[merged['receiving_city'].isna()]
nan_receiving_city['receiving_organization'].value_counts(dropna=False)

receiving_organization
IMT Atlantique Bretagne Pays de la Loire                 2
UNIVERSITAET KLAGENFURT                                  1
COPS GesmbH                                              1
Howard Duff                                              1
FORMA Design Bureau                                      1
ECOLE NATIONALE SUPERIEURE DES MINES D'ALES              1
DIMOKRITIO PANEPISTIMIO THRAKIS                          1
Ecole Nationale Supérieure des Mines de Saint-Etienne    1
Name: count, dtype: int64

To impute `receiving_city`, we use the city of the `receiving_organization`, easily obtained through an online search.

In [29]:
merged.loc[merged['receiving_organization'].eq('IMT Atlantique Bretagne Pays de la Loire'), 'receiving_city'] = 'Nantes'
merged.loc[merged['receiving_organization'].eq('UNIVERSITAET KLAGENFURT'), 'receiving_city'] = 'Klagenfurt'
merged.loc[merged['receiving_organization'].eq('COPS GesmbH'), 'receiving_city'] = 'Prague'
merged.loc[merged['receiving_organization'].eq('Howard Duff'), 'receiving_city'] = 'Unknown'
merged.loc[merged['receiving_organization'].eq('FORMA Design Bureau'), 'receiving_city'] = 'Ruse'
merged.loc[merged['receiving_organization'].eq("ECOLE NATIONALE SUPERIEURE DES MINES D'ALES"), 'receiving_city'] = 'Ales'
merged.loc[merged['receiving_organization'].eq('DIMOKRITIO PANEPISTIMIO THRAKIS'), 'receiving_city'] = 'Komotini'
merged.loc[merged['receiving_organization'].eq('Ecole Nationale Supérieure des Mines de Saint-Etienne'), 'receiving_city'] = 'Saint-Etienne'                                               

In [30]:
# Review remaining missing values.
missing = pd.concat([merged.isnull().sum(), 100 * merged.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing.sort_values(by='count', ascending=False)

count         %
receiving_organisation_erasmus_code      9  0.000260
receiving_organization                   4  0.000116
eu_grant_award_in_euros                  2  0.000058
sending_country_code                     0  0.000000
project_identifier                       0  0.000000
call_year                                0  0.000000
action_type                              0  0.000000
key_action                               0  0.000000
participants                             0  0.000000
receiving_city                           0  0.000000
receiving_country_code                   0  0.000000
sending_organisation_erasmus_code        0  0.000000
sending_organization                     0  0.000000
sending_city                             0  0.000000
project_reference                        0  0.000000
academic_year                            0  0.000000
groupleader                              0  0.000000
fewer_opportunities                      0  0.000000
special_needs                            0  0.000000
participant_profile                      0  0.000000
participant_gender                       0  0.000000
education_level                          0  0.000000
participant_nationality                  0  0.000000
field_of_education                       0  0.000000
activity_(mob)                           0  0.000000
mobility_duration                        0  0.000000
mobility_end_month                       0  0.000000
mobility_start_month                     0  0.000000
participant_age                          0  0.000000

`sending_organization` and `receiving_organization` were helpful in imputing data for other columns but they will not be used in our analysis; `sending_organisation_erasmus_code` and `receiving_organisation_erasmus_code` will also not be used. Additionally, our current dataset was joined on `project_identifier`so this duplicate of `project reference` can be dropped. We remove these five columns from our dataset.

In [31]:
merged = merged.drop(columns=['sending_organization', 'sending_organisation_erasmus_code', 'receiving_organization', 'receiving_organisation_erasmus_code', 'project_identifier'])

In [32]:
merged.shape

(3462076, 24)

In [33]:
# Review remaining missing values.
missing = pd.concat([merged.isnull().sum(), 100 * merged.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing.sort_values(by='count', ascending=False)

count         %
eu_grant_award_in_euros      2  0.000058
academic_year                0  0.000000
call_year                    0  0.000000
action_type                  0  0.000000
key_action                   0  0.000000
participants                 0  0.000000
receiving_city               0  0.000000
receiving_country_code       0  0.000000
sending_city                 0  0.000000
sending_country_code         0  0.000000
participant_age              0  0.000000
groupleader                  0  0.000000
project_reference            0  0.000000
special_needs                0  0.000000
participant_profile          0  0.000000
participant_gender           0  0.000000
education_level              0  0.000000
participant_nationality      0  0.000000
field_of_education           0  0.000000
activity_(mob)               0  0.000000
mobility_duration            0  0.000000
mobility_end_month           0  0.000000
mobility_start_month         0  0.000000
fewer_opportunities          0  0.000000

Finally, `eu_grant_award_in_euros` will be our target feature. Since there is no grant award information in these rows, we will remove them.

In [34]:
# Review rows without grant amount information.
merged[merged['eu_grant_award_in_euros'].isna()]

project_reference academic_year mobility_start_month  \
3402581  2019-1-PT01-KA101-060535     2019-2020              2019-11   
3402582  2019-1-PT01-KA101-060535     2019-2020              2019-11   

        mobility_end_month  mobility_duration  \
3402581            2019-11                7.0   
3402582            2019-11                7.0   

                             activity_(mob)       field_of_education  \
3402581  Structured Courses/Training Events  History and archaeology   
3402582  Structured Courses/Training Events                Languages   

        participant_nationality    education_level participant_gender  ...  \
3402581                      PT  ??? - ? Unknown ?               Male  ...   
3402582                      PT  ??? - ? Unknown ?             Female  ...   

        participant_age sending_country_code sending_city  \
3402581              42                   PT   Campanário   
3402582              48                   PT   Campanário   

        receiving_country_code receiving_city participants  \
3402581                     SI      Ljubljana          1.0   
3402582                     SI      Ljubljana          1.0   

                               key_action                      action_type  \
3402581  Learning Mobility of Individuals  School education staff mobility   
3402582  Learning Mobility of Individuals  School education staff mobility   

        call_year  eu_grant_award_in_euros  
3402581      2019                      NaN  
3402582      2019                      NaN  

[2 rows x 24 columns]

In [35]:
# Drop the NaN rows.
merged = merged.dropna(subset=['eu_grant_award_in_euros'])

In [36]:
# Review remaining missing values.
missing = pd.concat([merged.isnull().sum(), 100 * merged.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing.sort_values(by='count', ascending=False)

count    %
project_reference            0  0.0
academic_year                0  0.0
call_year                    0  0.0
action_type                  0  0.0
key_action                   0  0.0
participants                 0  0.0
receiving_city               0  0.0
receiving_country_code       0  0.0
sending_city                 0  0.0
sending_country_code         0  0.0
participant_age              0  0.0
groupleader                  0  0.0
fewer_opportunities          0  0.0
special_needs                0  0.0
participant_profile          0  0.0
participant_gender           0  0.0
education_level              0  0.0
participant_nationality      0  0.0
field_of_education           0  0.0
activity_(mob)               0  0.0
mobility_duration            0  0.0
mobility_end_month           0  0.0
mobility_start_month         0  0.0
eu_grant_award_in_euros      0  0.0

Now we review `education_level`. Approximately 1/3 of data is unknown.

In [37]:
merged['education_level'].value_counts()

education_level
??? - ? Unknown ?                                                                        1253797
ISCED-6 - First cycle / Bachelor’s or equivalent level (EQF-6)                           1204250
ISCED-7 - Second cycle / Master’s or equivalent level (EQF-7)                             606554
ISCED-3 - Upper secondary education                                                       175158
ISCED-5 - Short-cycle within the first cycle / Short-cycle tertiary education (EQF-5)      63551
ISCED-4 - Post-secondary non-tertiary education                                            55980
ISCED-8 - Third cycle / Doctoral or equivalent level (EQF-8)                               54461
ISCED-9 - Not elsewhere classified                                                         28011
ISCED-2 - Lower secondary education                                                        20312
Name: count, dtype: int64

To simplify, we replace '??? - ? Unknown ?' with 'Unknown'.

In [38]:
# Impute 'Unknown' and confirm.
merged['education_level'] = merged['education_level'].replace('??? - ? Unknown ?', 'Unknown')
merged['education_level'].value_counts()

education_level
Unknown                                                                                  1253797
ISCED-6 - First cycle / Bachelor’s or equivalent level (EQF-6)                           1204250
ISCED-7 - Second cycle / Master’s or equivalent level (EQF-7)                             606554
ISCED-3 - Upper secondary education                                                       175158
ISCED-5 - Short-cycle within the first cycle / Short-cycle tertiary education (EQF-5)      63551
ISCED-4 - Post-secondary non-tertiary education                                            55980
ISCED-8 - Third cycle / Doctoral or equivalent level (EQF-8)                               54461
ISCED-9 - Not elsewhere classified                                                         28011
ISCED-2 - Lower secondary education                                                        20312
Name: count, dtype: int64

Next, we review `participant_age` and see 346,225 entries of '-' for age, which is approximately 10% of our data.

In [39]:
merged['participant_age'].value_counts()

participant_age
21      422672
20      364280
22      352103
-       346225
23      291689
         ...  
819          1
-978         1
948          1
816          1
1040         1
Name: count, Length: 439, dtype: int64

We impute 0 for these unknown values and convert the column to an integer datatype.

In [40]:
# Impute 0 for '-' and convert column to an integer.
merged['participant_age'] = merged['participant_age'].replace('-', 0)
merged['participant_age'] = merged['participant_age'].astype(int)

We also review to see how many participant ages, including those just imputed to 0, are obvious outliers. 

In [41]:
# Calculate how many ages are less than 0 or greater than 100.
over_100 = merged['participant_age'][(merged['participant_age'] < 0) | (merged['participant_age'] > 100) | (merged['participant_age'] == 0)]
print(over_100) 

73           -1
165          -2
214          -1
229          -2
237          -1
           ... 
3448919       0
3450811       0
3454721    1040
3456223      -1
3457098       0
Name: participant_age, Length: 350699, dtype: int32


`participant_age` has 350,699 indisputable outliers and missing values. We will decide how to handle this column's outliers in the next section (EDA).

### Normalize Categorical Data <a id="categorical_data"></a>

We map `special_needs`, `fewer_opportunities`, and `groupleader` to Yes:1 and No:0, which will also convert these columns to an integer datatype.

In [42]:
# Function for imputing yes/no
def map_yes_no(df, col):
    df[col] = df[col].map({'Yes': 1, 'No': 0})
    return col

In [43]:
# Map 'special_needs' using function and confirm mapping.
map_yes_no(merged,'special_needs')
merged['special_needs'].value_counts(dropna=False)

special_needs
0    3439015
1      23059
Name: count, dtype: int64

In [44]:
# Map 'fewer_opportunities' using function and confirm mapping.
map_yes_no(merged, 'fewer_opportunities')
merged['fewer_opportunities'].value_counts(dropna=False)

fewer_opportunities
0    3244278
1     217796
Name: count, dtype: int64

In [45]:
# Map 'groupleader' using function and confirm mapping.
map_yes_no(merged,'groupleader')
merged['groupleader'].value_counts(dropna=False)

groupleader
0    3339759
1     122315
Name: count, dtype: int64

We map `participant_profile` to Learner:1 and Staff:0 and convert to integer.

In [46]:
# Map `participant_profile` Learner:1 and Staff:0, and confirm mapping.
merged['participant_profile'] = merged['participant_profile'].map({'Learner':1, 'Staff':0})
merged['participant_profile'].value_counts(dropna=False)

participant_profile
1    2667235
0     794839
Name: count, dtype: int64

### Convert Data Types <a id="convert_data"></a>

In [47]:
merged.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3462074 entries, 0 to 3462105
Data columns (total 24 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   project_reference        object 
 1   academic_year            object 
 2   mobility_start_month     object 
 3   mobility_end_month       object 
 4   mobility_duration        float64
 5   activity_(mob)           object 
 6   field_of_education       object 
 7   participant_nationality  object 
 8   education_level          object 
 9   participant_gender       object 
 10  participant_profile      int64  
 11  special_needs            int64  
 12  fewer_opportunities      int64  
 13  groupleader              int64  
 14  participant_age          int32  
 15  sending_country_code     object 
 16  sending_city             object 
 17  receiving_country_code   object 
 18  receiving_city           object 
 19  participants             float64
 20  key_action               object 
 21  action_type  

We convert `mobility_duration`, which is represented in calendar days, to an integer. We see that the project with the longest duration lasted approximately 18 months.

In [48]:
merged['mobility_duration'] = merged['mobility_duration'].astype(int)
merged['mobility_duration'].value_counts()

mobility_duration
5      305523
7      199130
8      159687
6      132415
12      91938
        ...  
416         1
414         1
398         1
433         1
577         1
Name: count, Length: 481, dtype: int64

We convert `call_year`, the year the grant was awarded, to an integer.

In [49]:
# Convert 'call_year'.
merged['call_year'] = merged['call_year'].astype(int)
merged['call_year'].dtype

dtype('int32')

### Review Date Columns <a id="review_dates"></a>

For `mobility_start_month`, we first confirm that all dates are between 2014-2019.

In [50]:
merged['mobility_start_month'].value_counts(dropna=False).sort_index()

mobility_start_month
2014-05         1
2014-06      3428
2014-07     10604
2014-08     34549
2014-09    124457
            ...  
2019-08     57024
2019-09    165642
2019-10     66242
2019-11     36479
2019-12     13655
Name: count, Length: 68, dtype: int64

We also confirm that all months lie between 1-12.

In [51]:
matches_start_month = merged['mobility_start_month'].str.extract(r'(-.*)')
matches_start_month.value_counts().sort_index()

-01    225162
-02    313376
-03    249008
-04    250826
-05    228429
-06    203133
-07    227572
-08    320135
-09    882726
-10    338305
-11    162908
-12     60494
Name: count, dtype: int64

Next, we review project end dates and find that some extend into 2021, which is within reason.

In [52]:
merged['mobility_end_month'].value_counts(dropna=False).sort_index()

mobility_end_month
2014-06      534
2014-07     5459
2014-08    13145
2014-09    12270
2014-10    17735
           ...  
2020-12       26
2021-01        9
2021-02       15
2021-03        7
2021-04        2
Name: count, Length: 83, dtype: int64

We then confirm that all months lie between 1-12.

In [53]:
matches_end_month = merged['mobility_end_month'].str.extract(r'(-.*)')
matches_end_month.value_counts().sort_index()

-01    304416
-02    298228
-03    199795
-04    231776
-05    353097
-06    537157
-07    404616
-08    231456
-09    248125
-10    201268
-11    175497
-12    276643
Name: count, dtype: int64

We review `call_year`.

In [54]:
merged['call_year'].value_counts()

call_year
2017    698173
2018    669909
2016    644429
2015    605519
2014    564463
2019    279581
Name: count, dtype: int64

### Remove Duplicates <a id="remove_duplicates"></a>

In [55]:
# Find duplicate rows across all columns.
merged.drop_duplicates()

project_reference academic_year mobility_start_month  \
0        2014-1-AT01-KA101-000059     2014-2015              2014-07   
1        2014-1-AT01-KA101-000059     2014-2015              2014-07   
2        2014-1-AT01-KA101-000059     2014-2015              2014-07   
3        2014-1-AT01-KA101-000059     2014-2015              2014-08   
4        2014-1-AT01-KA101-000059     2015-2016              2015-04   
...                           ...           ...                  ...   
3462101  2019-2-UK01-KA105-062363     2019-2020              2019-09   
3462102  2019-2-UK01-KA105-062363     2019-2020              2019-09   
3462103  2019-2-UK01-KA105-062363     2019-2020              2019-09   
3462104  2019-2-UK01-KA105-062363     2019-2020              2019-09   
3462105  2019-2-UK01-KA105-062363     2019-2020              2019-09   

        mobility_end_month  mobility_duration  \
0                  2014-07                 12   
1                  2014-08                  7   
2                  2014-08                 12   
3                  2014-08                 12   
4                  2015-04                  4   
...                    ...                ...   
3462101            2019-10                  7   
3462102            2019-10                  7   
3462103            2019-10                  7   
3462104            2019-10                  7   
3462105            2019-10                  7   

                                          activity_(mob)  \
0                                  Staff training abroad   
1                                  Staff training abroad   
2                                  Staff training abroad   
3                                  Staff training abroad   
4                                  Staff training abroad   
...                                                  ...   
3462101  Mobility of youth workers - Programme Countries   
3462102  Mobility of youth workers - Programme Countries   
3462103  Mobility of youth workers - Programme Countries   
3462104  Mobility of youth workers - Programme Countries   
3462105  Mobility of youth workers - Programme Countries   

           field_of_education participant_nationality education_level  \
0        Language acquisition                      AT         Unknown   
1        Language acquisition                      AT         Unknown   
2        Language acquisition                      AT         Unknown   
3        Language acquisition                      CZ         Unknown   
4        Language acquisition                      AT         Unknown   
...                       ...                     ...             ...   
3462101           ? Unknown ?                      TR         Unknown   
3462102           ? Unknown ?                      UK         Unknown   
3462103           ? Unknown ?                      UK         Unknown   
3462104           ? Unknown ?                      UK         Unknown   
3462105           ? Unknown ?                      UK         Unknown   

        participant_gender  ...  participant_age  sending_country_code  \
0                   Female  ...               42                    AT   
1                   Female  ...               25                    AT   
2                   Female  ...               51                    AT   
3                   Female  ...               34                    AT   
4                   Female  ...               61                    AT   
...                    ...  ...              ...                   ...   
3462101             Female  ...               24                    UK   
3462102             Female  ...               32                    UK   
3462103             Female  ...               37                    UK   
3462104               Male  ...               41                    UK   
3462105               Male  ...               37                    UK   

         sending_city  receiving_country_code  receiving_city participants 

In [56]:
merged.shape

(3462074, 24)

### Rename Columns <a id="rename_columns"></a>

In [57]:
# Rename columns
merged.rename(columns = {'project_reference':'project_id', 'mobility_start_month': 'start_month', 'mobility_end_month': 'end_month', 'mobility_duration':'duration_in_days', 'activity_(mob)':'activity', 'participant_profile':'participant_role', 'participants': '# of participants', 'sending_country_code':'sending_country', 'receiving_country_code':'receiving_country', 'eu_grant_award_in_euros':'grant_in_euros'}, inplace=True)

In [58]:
# Display the first few rows of the dataset
merged.head().T

0  \
project_id                       2014-1-AT01-KA101-000059   
academic_year                                   2014-2015   
start_month                                       2014-07   
end_month                                         2014-07   
duration_in_days                                       12   
activity                            Staff training abroad   
field_of_education                   Language acquisition   
participant_nationality                                AT   
education_level                                   Unknown   
participant_gender                                 Female   
participant_role                                        0   
special_needs                                           0   
fewer_opportunities                                     0   
groupleader                                             0   
participant_age                                        42   
sending_country                                        AT   
sending_city                                         Wels   
receiving_country                                      ES   
receiving_city                                     Malaga   
# of participants                                     1.0   
key_action               Learning Mobility of Individuals   
action_type               School education staff mobility   
call_year                                            2014   
grant_in_euros                                    17205.0   

                                                        1  \
project_id                       2014-1-AT01-KA101-000059   
academic_year                                   2014-2015   
start_month                                       2014-07   
end_month                                         2014-08   
duration_in_days                                        7   
activity                            Staff training abroad   
field_of_education                   Language acquisition   
participant_nationality                                AT   
education_level                                   Unknown   
participant_gender                                 Female   
participant_role                                        0   
special_needs                                           0   
fewer_opportunities                                     0   
groupleader                                             0   
participant_age                                        25   
sending_country                                        AT   
sending_city                                         Wels   
receiving_country                                      UK   
receiving_city                                 Canterbury   
# of participants                                     1.0   
key_action               Learning Mobility of Individuals   
action_type               School education staff mobility   
call_year                                            2014   
grant_in_euros                                    17205.0   

                                                        2  \
project_id                       2014-1-AT01-KA101-000059   
academic_year                                   2014-2015   
start_month                                       2014-07   
end_month                                         2014-08   
duration_in_days                                       12   
activity                            Staff training abroad   
field_of_education                   Language acquisition   
participant_nationality                                AT   
education_level                                   Unknown   
participant_gender                                 Female   
participant_role                                        0   
special_needs                                           0   
fewer_opportunities                                     0   
groupleader                                             0   
participant_age                                        51   
sending_country             